### ISSUE TITLE
resnet+yolo3head where should I put the route layer 

### ISSUE THREAD

**Issue by springshuai** [AlexeyAB/darknet#2005]
Hi AlexeyAB, 

Thanks for you repo. I managed to run the yolo3-voc but since the darknet is quite large so I am trying to replace it by ResNet50. 
My question is how should I write the route layers? E.g. in yolo3 I found [route] layers = -1, 61 and [route] layers = -1, 36. And in darknet53.cfg I found that the 36th (and the 61th respectively) layer --- if my counting is correct --- is right before the 11th (and the 19th respectively) [shortcut] layer. I am confused cause I thought the concatenate layer is attached AFTER the shortcuts, no? If I want to connect the yolo3 head on a ResNet50 backbone, what are the layer numbers for the [route] layers?

Thanks a lot.

**Comment by AlexeyAB** [comment]
@springshuai Hi,

>  I thought the concatenate layer is attached AFTER the shortcuts

What do you mean?


You should connect to the layers 53 and 29, because these [shortcut] layers have size 2x (16x16) and 4x (32x32) more than the final size (8x8) respectively, i.e. they are located before subsampling (stride=2) layers.
ResNet50:
![image](https://user-images.githubusercontent.com/4096485/49345350-a34b1680-f694-11e8-98bf-0e9ae269fcdb.png)

**Comment by AlexeyAB** [comment]
@springshuai 

> BUT, the network structure (http://ethereon.github.io/netscope/#/gist/db945b393d40bfa26006) I found here (https://github.com/KaimingHe/deep-residual-networks) do the subsampling in those 1x1 convolutional layers (e.g. layer 30 and 54).

It doesn't matter. Current versions of Resnet 50 has good accuracy 75.8 Top1 | 92.9 Top5: https://pjreddie.com/darknet/imagenet/

So just copy structure from: https://github.com/AlexeyAB/darknet/blob/master/cfg/resnet50.cfg

----

> So my question is how can I use the pertained weights of ResNet50 for my ResNet50+YoloV3 model?

* Download this weights: https://pjreddie.com/media/files/resnet50.weights

* Do this command: https://github.com/AlexeyAB/darknet/blob/21a4ec9390b61c0baa7ef72e72e59fa143daba4c/build/darknet/x64/partial.cmd#L36
`./darknet partial cfg/resnet50.cfg resnet50.weights resnet50.65 65`

* And run training using `resnet50.65` pre-trained file:
`./darknet detector train data/obj.data yolov3-resnet50.cfg resnet50.65`

**Comment by springshuai** [comment]
Huuuge thanks to you. Since the training went well :-), here I have more question awaiting:

1-The mAP of my resne50t(with pertained weights and input resolution is **416x416**) + yolov3 is around **0.43** after 60 epochs. It is obviously way behind the original darknet53-based one, which is **0.55** according to the author. I wasn't expecting this since resnet50's Top-1 accuracy on ImageNet is only **1.4** percents behind darknet53. Do you think, from your experience, such a performance gap is normal? 

2-The yolov3 head is pretty heavy, e.g., there is a 3x3 convolution between featuremap_h x featuremap_w x 512 and featuremap_h x featuremap_w x 1024, thus parameters will be **3x3x512x1024**. So I am thinking of using **depth-wise + 1x1 convolution** to reduce the size. Do you think this will bring a significant performance drop? And it this modification possible in the .cfg file?

Thanks in advance.

**Comment by springshuai** [comment]
@aftaufik Hi, in my case, I copied the resnet50.cfg, replaced its classification head by YOLOV3 detection head (make sure the route layers are correctly adapted), changed the learning rate and left the rest part of resnet50.cfg untouched.
